[export] Modify SDPA decomposition to decompose _scaled_dot_product_flash_attention_for_cpu

larryliu0820 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117097

Summary: As titled. #115913 added
`_scaled_dot_product_flash_attention_for_cpu` and the export result of
`scaled_dot_product_attention` includes this op. Adding this
decomposition so that it's being decomposed the same way as
`_scaled_dot_product_attention_math`.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

larryliu0820 committed:
[export] Modify SDPA decomposition to decompose _scaled_dot_product_flash_attention_for_cpu

Summary: As titled. #115913 added
`_scaled_dot_product_flash_attention_for_cpu` and the export result of
`scaled_dot_product_attention` includes this op. Adding this
decomposition so that it's being decomposed the same way as
`_scaled_dot_product_attention_math`.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

larryliu0820 committed:
Update on "[export] Modify SDPA decomposition to decompose _scaled_dot_product_flash_attention_for_cpu"


Summary: As titled. #115913 added
`_scaled_dot_product_flash_attention_for_cpu` and the export result of
`scaled_dot_product_attention` includes this op. Adding this
decomposition so that it's being decomposed the same way as
`_scaled_dot_product_attention_math`.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

lezcano reviewed (commented):
Don't we want to keep the decomp of `_scaled_dot_product_flash_attention`?

kimishpatel commented:
I have been wanting to fix this in a more correct way which probably is non-trivial. Correct way is really to define decomp for aten sdpa op. I remember discussing this with you, but I still think we should look into decomposing aten sdpa.

One not so nice executorch specific workaround is inside to_edge, where before calling run_decomposition, we manually replace instances of aten sdpa with its decomp version.

kimishpatel commented:
> Don't we want to keep the decomp of `_scaled_dot_product_flash_attention`?

why flash attention?

lezcano commented:
This PR removes one decomposition and registers a decomposition in one of the functions inside it. Why don't we keep the original decomposition in terms of this new decomposition?

kimishpatel commented:
> This PR removes one decomposition and registers a decomposition in one of the functions inside it. Why don't we keep the original decomposition in terms of this new decomposition?

Oh I see. Sorry didnt follow the first time around. Yeah makes sense. Basically you are suggesting to keep definition for both decomp and one of it just calls the other one. @larryliu0820 ?

lezcano commented:
Yep, although looking at the removed code, the previous decomposition was completely wrong, with lines like
```python
logsumexp = torch.empty([batchSize, qSize, num_head, headSize], dtype=torch.float)
```
so it may be alright to straight up remove it.

larryliu0820 commented:
> Yep, although looking at the removed code, the previous decomposition was completely wrong, with variables like
> 
> ```python
> logsumexp = torch.empty([batchSize, qSize, num_head, headSize], dtype=torch.float)
> ```
> 
> that are never filled up, so it may be alright to straight up remove it.

Yeah the previous decomposition was pretty awkward because `_scaled_dot_product_flash_attention` is returning way more things than `_scaled_dot_product_attention_math`. The current decomp makes much more sense.

lezcano commented on test/test_decomp.py:
```diff
@@ -927,7 +927,7 @@ def forward(self, query_layer, key_layer, value_layer):
             decomposition_table=get_decompositions(
                 [
-                    torch.ops.aten._scaled_dot_product_flash_attention.default,
+                    torch.ops.aten._scaled_dot_product_flash_attention_for_cpu.default,
```
I guess this test is not meaningful for CUDA any more.

larryliu0820 commented:
@pytorchbot merge

lezcano commented:
@larryliu0820 can you please deactivate the test for CUDA?

larryliu0820 commented:
> @larryliu0820 can you please deactivate the test for CUDA?

oh sorry I'm on it

larryliu0820 committed:
Update on "[export] Modify SDPA decomposition to decompose _scaled_dot_product_flash_attention_for_cpu"


Summary: As titled. #115913 added
`_scaled_dot_product_flash_attention_for_cpu` and the export result of
`scaled_dot_product_attention` includes this op. Adding this
decomposition so that it's being decomposed the same way as
`_scaled_dot_product_attention_math`.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

larryliu0820 commented:
@pytorchbot merge

larryliu0820 commented:
@pytorchbot merge
